### ISSUE TITLE
`InitializationOptions` and `ClientExperimentalCapablities` can result in a null pointer exceptions

### ISSUE THREAD

**Issue by ckipp01** [scalameta/metals#1668]
**Describe the bug**
We haven't hit this yet because both `coc-metals` and the VS Code extension set all the `ClientExperimentalCapabilities`. However, if you remove one of them, that value will become `null` in the case class representing it. The same happens for `InitializationOptions`. When there are no `InitializationOptions`, some clients will not even send in the key. This will result in the `IntializationOptions.Default`, which is fine. However, if the client sends in an empty `{}`, then the result will be the following:

```
InitializationOptions.scala:37 x: InitializationOptions(null, null, null, null, null, null, null, null, null, null)
```

Then as you can guess, when we check for these values, it results in a null pointer exception.

**To Reproduce**
1. Remove a value in either `coc-metals` or the VS Code exception that we do a check against or use a client that sends an empty object for either of these.

**Expected behavior**
If the key doesn't exist, the default value should get applied. It shouldn't be an all or nothing.

You can see it's an all or nothing here:

https://github.com/scalameta/metals/blob/810a09c5584642ba7a5196d538342607615b88cb/metals/src/main/scala/scala/meta/internal/metals/ClientExperimentalCapabilities.scala#L33-L34

And also here:

https://github.com/scalameta/metals/blob/810a09c5584642ba7a5196d538342607615b88cb/metals/src/main/scala/scala/meta/internal/metals/InitializationOptions.scala#L34-L35

**Search terms**
clientExperimentalCapabilities, initializationOptions

**Comment by ayoub-benali** [comment]
@ckipp01 I don't know how easy it would be to handle missing keys or values in Gson lib but I think these settings should also changed to `Option` with default `None` to have better semantic. 
So that when merging values from server properties, InitializationOptions, ClientExperimentalCapablities we have better context.

**Comment by dangerousben** [comment]
This affects emacs lsp-mode, given the following (trimmed) init request:

```
{
  "processId": null,
  "rootPath": "/home/ben/src/typelevel/cats",
  "rootUri": "file:///home/ben/src/typelevel/cats",
  "capabilities": {
    "experimental": {
      "decorationProvider": true,
      "debuggingProvider": true,
      "didFocusProvider": true,
      "executeClientCommandProvider": true,
      "doctorProvider": "html",
      "statusBarProvider": "on"
    }
  },
  "clientInfo": {
    "name": "emacs",
    "version": "GNU Emacs 28.0.50 (build 51, x86_64-pc-linux-gnu, GTK+ Version 3.24.12, cairo version 1.16.0)\n of 2020-04-28"
  }
}
```

There is no `initializationOptions` key, but this results in a `JNull` from `initializeParams.getInitializationOptions()` and so fails in the same way as if the key were present with an empty object.

The resulting exception is:
```
java.lang.NullPointerException at scala.meta.internal.metals.ConfiguredLanguageClient.logMessage(ConfiguredLanguageClient.scala:99)
```
where line 99 is currently:
```scala
      initializationOptions.statusBarIsLogMessage
```

**Comment by ayoub-benali** [comment]
> So one thing that doesn't make a lot of sense in this situation is that coc-metals also doesn't set any initializationOptions and leaves out the key altogether, just like in your case

I think because often we do something like 
```scala
    val statusBarIsShow = config.statusBar.isShowMessage ||
      clientCapabilities.statusBarIsShowMessage ||
      initializationOptions.statusBarIsShowMessage
```
with `coc-metals` `clientCapabilities.statusBarIsShowMessage` is set to true so we exit early and never check for `initializationOptions` which has value null.

For the current issue it is only due to surprise from Gson lib, parsing tests could have prevented that. 

But in general supporting only one way to enable settings will make life easier yes but at the cost of clients, they would have to update their internals in order to upgrade Metals version.

**Comment by ckipp01** [comment]
> with coc-metals clientCapabilities.statusBarIsShowMessage is set to true so we exit early and never check for initializationOptions which has value null.

But that should be the _exact_ same situation that @dangerousben is having. Neither of us have `clientCapabilities` being sent in. So our `clientCapabilities` should be identical. I just tested this locally and set my exactly `clientExperimentalCapabilities` to what they had in the pastebin and I'm unable to mimic it for some reason.

**Comment by ckipp01** [comment]
Closed via #1674
